### PR TITLE
spark support yarn cluster

### PR DESCRIPTION
**Pull request by ChengJie1053**
spark support yarn cluster

### CODE REVIEW

File: linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/AbstractEngineConnLaunchService.scala [hunk 1]
```diff
@@ -146,11 +148,19 @@ abstract class AbstractEngineConnLaunchService extends EngineConnLaunchService w
       throw t
     }
     LoggerUtils.removeJobIdMDC()
+    val deployMode: String =
+      request.creationDesc.properties.getOrDefault("spark.submit.deployMode", "client")
```
Comment (peacewong) [inline linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/AbstractEngineConnLaunchService.scala:152]: should judge  by yarn cluster mode label

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnFactory.scala [hunk 2]
```diff
@@ -144,19 +144,33 @@ class SparkEngineConnFactory extends MultiExecutorEngineConnFactory with Logging
...
-      val sparkConfValue1 = Utils.tryQuietly(CommonVars("spark.yarn.dist.files", "").getValue)
-      val sparkConfValue2 = Utils.tryQuietly(sparkConf.get("spark.yarn.dist.files"))
-      if (StringUtils.isNotBlank(sparkConfValue2)) {
-        pythonLibUris = sparkConfValue2 +: pythonLibUris
-      }
-      if (StringUtils.isNotBlank(sparkConfValue1)) {
-        pythonLibUris = sparkConfValue1 +: pythonLibUris
+
+    // Set deploy-mode with the optional values `cluster`and `client`, the default value `client`
+    val deployMode: String = SPARK_DEPLOY_MODE.getValue(options)
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnFactory.scala:149]: should to use label judge

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala [hunk 3]
```diff
@@ -61,7 +66,7 @@ class SparkSubmitProcessEngineConnLaunchBuilder(builder: JavaProcessEngineConnLa
     val executorMemory = getValueAndRemove(properties, LINKIS_SPARK_EXECUTOR_MEMORY)
     val numExecutors = getValueAndRemove(properties, LINKIS_SPARK_EXECUTOR_INSTANCES)
 
-    val files = getValueAndRemove(properties, "files", "").split(",").filter(isNotBlankPath)
+    var files = getValueAndRemove(properties, "files", "").split(",").filter(isNotBlankPath)
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala:69]: update to array buffer

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/impl/DefaultEngineConnPidCallbackService.java [hunk 4]
```diff
@@ -56,6 +63,13 @@ public void dealPid(ResponseEngineConnPid protocol) {
     }
 
     engineNode.setIdentifier(protocol.pid());
+
+    if (engineNode.getMark().equals(AMConstant.CLUSTER_PROCESS_MARK)) {
+      ServiceInstance serviceInstance = protocol.serviceInstance();
```
Comment (peacewong) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/am/service/impl/DefaultEngineConnPidCallbackService.java:68]: You should first org.apache.linkis.manager.am.manager.EngineNodeManager#updateEngineNode and then update label

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/label/service/impl/DefaultNodeLabelService.java [hunk 5]
```diff
@@ -196,6 +196,84 @@ public void updateLabelsToNode(ServiceInstance instance, List<Label<?>> labels)
...
+  public void labelsFromInstanceToNewInstance(
+      ServiceInstance oldServiceInstance, ServiceInstance newServiceInstance) {
+    List<PersistenceLabel> labels =
+        labelManagerPersistence.getLabelByServiceInstance(newServiceInstance);
+    List<String> newKeyList = labels.stream().map(Label::getLabelKey).collect(Collectors.toList());
+    List<PersistenceLabel> nodeLabels =
+        labelManagerPersistence.getLabelByServiceInstance(oldServiceInstance);
+
+    List<String> oldKeyList =
+        nodeLabels.stream().map(InheritableLabel::getLabelKey).collect(Collectors.toList());
```
Comment (peacewong) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/label/service/impl/DefaultNodeLabelService.java:209]: It shouldn’t be so complicated here, we only need to first get the new instance tag, then get the old instance tag, then delete the duplicate tag in the old tag, and then associate the new instance with the old tag just go
这里应该不用搞这么复杂，我们只需要第一步获得新的instance的标签，然后再获取老的instance标签，然后将老的标签中重复的标签删除，接着再将新的instance和老的标签关联上就好了

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala [hunk 6]
```diff
@@ -115,8 +124,32 @@ class SparkSubmitProcessEngineConnLaunchBuilder(builder: JavaProcessEngineConnLa
       memory
     }
 
+    //    val deployMode: String = getValueAndRemove(properties, SPARK_DEPLOY_MODE)
+    val deployMode: String = SPARK_DEPLOY_MODE.getValue(properties)
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala:128]: Should use this method ：
 val label = LabelUtil.getEngingeConnRuntimeModeLabel(engineCreationContext.getLabels())
    val isYarnClusterMode: Boolean =
      if (null != label && label.getModeValue.equals(LabelValueConstant.YARN_CLUSTER_VALUE)) true
      else false

File: linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/AbstractEngineConnLaunchService.scala (hunk 1 above)
Comment (peacewong) [reply linkis-computation-governance/linkis-engineconn-manager/linkis-engineconn-manager-server/src/main/scala/org/apache/linkis/ecm/server/service/impl/AbstractEngineConnLaunchService.scala:153]: Should use this method ：
val label = LabelUtil.getEngingeConnRuntimeModeLabel(engineCreationContext.getLabels())
val isYarnClusterMode: Boolean =
if (null != label && label.getModeValue.equals(LabelValueConstant.YARN_CLUSTER_VALUE)) true
else false

can add this method to LabelUtil class

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala (hunk 6 above)
Comment (ChengJie1053) [reply linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/launch/SparkSubmitProcessEngineConnLaunchBuilder.scala:128]: Ok, thanks for reviewing the code

**Pull request merged**